File: src/main/java/org/academy/internal/client/gui/screen/WirelessNodeScreen.java
package org.academy.internal.client.gui.screen;

import net.minecraft.client.Minecraft;
import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;
import net.minecraft.world.entity.player.Inventory;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.neoforge.common.NeoForge;
import org.academy.api.client.Resource;
import org.academy.api.client.gui.animation.EasingFunctions;
import org.academy.api.client.gui.animation.ObjectAnimator;
import org.academy.api.client.gui.layout.Gravity;
import org.academy.api.client.gui.layout.Orientation;
import org.academy.api.client.gui.layout.SizeMode;
import org.academy.api.client.gui.msdf.atlas.MsdfAtlasDebugger;
import org.academy.api.client.gui.msdf.atlas.MsdfAtlasManager;
import org.academy.api.client.gui.msdf.font.MsdfFontService;
import org.academy.api.client.gui.screen.ContainerUiScreen;
import org.academy.api.client.gui.util.InfoAreaUtil;
import org.academy.api.client.gui.util.WirelessPanelUtil;
import org.academy.api.client.gui.widget.*;
import org.academy.api.client.util.AnimationUtil;
import org.academy.api.common.wireless.SetNodeNamePacket;
import org.academy.api.common.wireless.SetNodePassPacket;
import org.academy.internal.common.world.inventory.WirelessNodeMenu;
import org.academy.internal.common.world.level.block.entity.WirelessNodeBlockEntity;
import org.jspecify.annotations.Nullable;
import org.misaka.MisakaNetworkClient;

import java.util.ArrayList;
import java.util.function.Consumer;

import static org.academy.api.client.gui.util.InfoAreaUtil.*;

public final class WirelessNodeScreen extends ContainerUiScreen<WirelessNodeMenu> {
    private final BlockPos mainPos;
    private final WirelessNodeBlockEntity wirelessNodeBlockEntity;
    private int ticks;
    private Consumer<String> energyValueSetter = _ -> {
    };
    private Consumer<String> capacityValueSetter = _ -> {
    };
    private Consumer<String> rangeValueSetter = _ -> {
    };

    public WirelessNodeScreen(WirelessNodeMenu menu, Inventory playerInventory, Component title, WirelessNodeBlockEntity blockEntity) {
        super(menu, playerInventory, title);
        mainPos = blockEntity.getBlockPos();

        wirelessNodeBlockEntity = blockEntity;
        NeoForge.EVENT_BUS.register(this);
    }

    @Override
    public void onClose() {
        super.onClose();
        NeoForge.EVENT_BUS.unregister(this);
    }

    @Nullable
    public static WirelessNodeScreen create(WirelessNodeMenu menu, Inventory playerInventory, Component title, BlockPos mainPos) {
        var set = new ArrayList<>(MsdfFontService.getInstance().getLoadedFonts().keySet());
        for (var i = 0; i < set.size(); i++) {
            MsdfAtlasDebugger.dumpAtlas(MsdfAtlasManager.getInstance().getAtlas(set.get(i)), "" + i);
        }

        if (Minecraft.getInstance().level != null && Minecraft.getInstance().level.getBlockEntity(mainPos) instanceof WirelessNodeBlockEntity blockEntity) {
            return new WirelessNodeScreen(menu, playerInventory, title, blockEntity);
        } else {
            return null;
        }
    }

    @Override
    protected void onInit(RadioGroupWidget pageButtons, RadioButtonWidget invButton, FrameLayoutWidget content, FrameLayoutWidget invPage) {
        var duration = 600L;
        var childDuration = duration - 100;

        var ui = new ImageWidget(Resource.Textures.WIRELESS_NODE_UI);
        ui.setLayoutParams(
                new FrameLayoutWidget.LayoutParams()
                        .sizeMode(SizeMode.MATCH_PARENT)
        );
        invPage.addChild("ui", ui);

        var effect = new SpriteSheetWidget(
                Resource.Textures.WIRELESS_NODE_STATE,
                Orientation.VERTICAL,
                186, 750,
                186, 75,
                10
        ) {
            @Override
            public void tick() {
                var progressCapacity = (float) wirelessNodeBlockEntity.connectedUsersCount / wirelessNodeBlockEntity.maxConnectedUsers;

                if (Float.isNaN(progressCapacity)) progressCapacity = 0;

                int index;
                if (wirelessNodeBlockEntity.connectedUsersCount == 0) {
                    index = (ticks / 20) % 2 == 0 ? 8 : 9;
                } else {
                    index = Math.max(0, Math.min((int) (progressCapacity * 8 - 1), 7));
                }

                setFrameIndex(index);
            }
        };
        effect.setLayoutParams(
                new FrameLayoutWidget.LayoutParams()
                        .heightMode(SizeMode.MATCH_PARENT)
                        .width(186 / 2f)
                        .gravity(Gravity.CENTER_HORIZONTAL)
                        .padding(0, 33.5f, 0, 116)
        );
        invPage.addChild("effect", effect);

        var wirelessPage = WirelessPanelUtil.create(mainPos, true);
        wirelessPage.setVisibility(Widget.Visibility.GONE);
        wirelessPage.setEnabled(false);
        content.addChild("page_wireless", wirelessPage);

        var wirelessButton = createButton(Resource.Textures.ICON_WIRELESS);
        wirelessButton.setLayoutParams(
                new WidgetContainer.LayoutParams()
                        .widthMode(SizeMode.MATCH_PARENT)
                        .height(16)
        );
        pageButtons.addChild("wireless", wirelessButton);
        pageButtons.setOnSelectionChanged(button -> {
            switch (button.getName()) {
                case "inv" -> {
                    AnimationUtil.hide(wirelessPage);
                    AnimationUtil.show(invPage);
                    setHandleContainer(true);
                    setRenderInventory(true);
                }
                case "wireless" -> {
                    AnimationUtil.hide(invPage);
                    AnimationUtil.show(wirelessPage);
                    setHandleContainer(false);
                    setRenderInventory(false);
                }
            }
        });
        pageButtons.selectButton(invButton);

        var info = InfoAreaUtil.create(this, leftPos + imageWidth, topPos - 22);
        {
            var energyValueLabel = new LabelWidget("0 AF");
            energyValueSetter = energyValueLabel::setText;
            var energyLayout = createInfoRow("ENERGY", "icon_energy", 0xFF25C4FF, energyValueLabel);
            info.addChild("energy_layout", energyLayout);

            var capacityValueLabel = new LabelWidget("0 / 0");
            capacityValueSetter = capacityValueLabel::setText;
            var capacityLayout = createInfoRow("CAPACITY", "icon_capacity", 0xFFFF6C00, capacityValueLabel);
            info.addChild("capacity_layout", capacityLayout);

            var infoLabel = new LabelWidget("Information");
            infoLabel.setLayoutParams(
                    new LinearLayoutWidget.LayoutParams()
                            .padding(6.5f, 0, 0, 0)
            );
            infoLabel.setScale(0.75f);
            info.addChild("label_info", infoLabel);

            var rangeValueLabel = new LabelWidget("0");
            rangeValueSetter = rangeValueLabel::setText;
            rangeValueLabel.setLayoutParams(
                    new WidgetContainer.LayoutParams()
                            .size(12, 12)
                            .gravity(Gravity.CENTER)
            );
            var rangeLayout = createAttributeRow("Trans. Range", rangeValueLabel, 52);
            info.addChild("range_layout", rangeLayout);

            var nameTextBox = new TextBoxWidget(12);
            nameTextBox.setBackground(null);
            nameTextBox.setWhenEnter(s -> MisakaNetworkClient.sendPacket(new SetNodeNamePacket(wirelessNodeBlockEntity.getBlockPos(), s)));
            var nameLayout = createAttributeRow("Node Name", createInputRow(nameTextBox), 48);
            info.addChild("name_layout", nameLayout);

            var passTextBox = new TextBoxWidget(12);
            passTextBox.setBackground(null);
            passTextBox.setWhenEnter(s -> MisakaNetworkClient.sendPacket(new SetNodePassPacket(wirelessNodeBlockEntity.getBlockPos(), s)));
            var passLayout = createAttributeRow("Password", createInputRow(passTextBox), 42);
            info.addChild("pass_layout", passLayout);
        }

        pageButtons.startAnimation(ObjectAnimator.ofFloat(pageButtons::setAlpha, 0f, 1f).setDuration(childDuration));
        pageButtons.startAnimation(ObjectAnimator.ofFloat(pageButtons::setTranslationY, 20, 0).setDuration(duration).setInterpolator(EasingFunctions.EASE_OUT_CUBIC));

        updateInfo();
    }

    private void updateInfo() {
        ticks++;

        capacityValueSetter.accept(wirelessNodeBlockEntity.connectedUsersCount + " / " + wirelessNodeBlockEntity.maxConnectedUsers);
        energyValueSetter.accept(WindGenScreen.AF.formatted(wirelessNodeBlockEntity.getEnergyStored()));
        rangeValueSetter.accept(wirelessNodeBlockEntity.radius + "");
    }

    @Override
    protected void containerTick() {
        super.containerTick();
        updateInfo();
    }

    @SubscribeEvent
    public void onFocusGainedEvent(TextBoxWidget.FocusGainedEvent event) {
        setHandleContainer(false);
    }

    @SubscribeEvent
    public void onFocusLostEvent(TextBoxWidget.FocusLostEvent event) {
        setHandleContainer(true);
    }
}

File: src/main/kotlin/org/academy/internal/client/gui/screen/WirelessNodeScreen.kt
package org.academy.internal.client.gui.screen

import net.minecraft.client.Minecraft
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.minecraft.world.entity.player.Inventory
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.neoforge.common.NeoForge
import org.academy.api.client.Resource
import org.academy.api.client.gui.animation.EasingFunctions
import org.academy.api.client.gui.animation.ObjectAnimator
import org.academy.api.client.gui.layout.Gravity
import org.academy.api.client.gui.layout.Orientation
import org.academy.api.client.gui.layout.SizeMode
import org.academy.api.client.gui.msdf.atlas.MsdfAtlasDebugger
import org.academy.api.client.gui.msdf.font.MsdfFontService
import org.academy.api.client.gui.screen.ContainerUiScreen
import org.academy.api.client.gui.util.InfoAreaUtil.create
import org.academy.api.client.gui.util.InfoAreaUtil.createAttributeRow
import org.academy.api.client.gui.util.InfoAreaUtil.createInfoRow
import org.academy.api.client.gui.util.InfoAreaUtil.createInputRow
import org.academy.api.client.gui.util.WirelessPanelUtil.create
import org.academy.api.client.gui.widget.*
import org.academy.api.client.gui.widget.TextBoxWidget.FocusGainedEvent
import org.academy.api.client.gui.widget.TextBoxWidget.FocusLostEvent
import org.academy.api.client.util.AnimationUtil
import org.academy.api.common.wireless.SetNodeNamePacket
import org.academy.api.common.wireless.SetNodePassPacket
import org.academy.internal.common.world.inventory.WirelessNodeMenu
import org.academy.internal.common.world.level.block.entity.WirelessNodeBlockEntity
import org.misaka.MisakaNetworkClient
import java.lang.Float
import java.util.function.Consumer
import kotlin.Int
import kotlin.String
import kotlin.run

class WirelessNodeScreen(
    menu: WirelessNodeMenu,
    playerInventory: Inventory,
    title: Component,
    private val wirelessNodeBlockEntity: WirelessNodeBlockEntity
) : ContainerUiScreen<WirelessNodeMenu>(menu, playerInventory, title) {
    private val mainPos: BlockPos = wirelessNodeBlockEntity.blockPos
    private var ticks = 0
    private var energyValueSetter =  { _: String -> }
    private var capacityValueSetter =  { _: String -> }
    private var rangeValueSetter =  { _: String -> }

    init {
        NeoForge.EVENT_BUS.register(this)
/*        for (msdfFont in MsdfFontService.loadedFonts.values) {
            MsdfAtlasDebugger.dumpAtlas(msdfFont.atlas,
                msdfFont.descriptor.identifier.path.length.toString()
            )
        }*/
    }

    override fun onClose() {
        super.onClose()
        NeoForge.EVENT_BUS.unregister(this)
    }

    override fun onInit(
        pageButtons: RadioGroupWidget,
        invButton: RadioButtonWidget,
        content: FrameLayoutWidget,
        invPage: FrameLayoutWidget
    ) {
        val duration = 600L
        val childDuration = duration - 100

        val ui = ImageWidget(Resource.Textures.WIRELESS_NODE_UI)
        ui.layoutParams = FrameLayoutWidget.LayoutParams()
            .sizeMode(SizeMode.MATCH_PARENT)

        invPage.addChild("ui", ui)

        val effect: SpriteSheetWidget = object : SpriteSheetWidget(
            Resource.Textures.WIRELESS_NODE_STATE,
            Orientation.VERTICAL,
            186, 750,
            186, 75,
            10
        ) {
            override fun tick() {
                var progressCapacity =
                    wirelessNodeBlockEntity.connectedUsersCount.toFloat() / wirelessNodeBlockEntity.maxConnectedUsers

                if (Float.isNaN(progressCapacity)) progressCapacity = 0f
                val index: Int = if (wirelessNodeBlockEntity.connectedUsersCount == 0) {
                    if ((ticks / 20) % 2 == 0) 8 else 9
                } else {
                    Math.clamp((progressCapacity * 8 - 1).toInt().toLong(), 0, 7)
                }

                frameIndex = index
            }
        }
        effect.layoutParams = FrameLayoutWidget.LayoutParams()
            .heightMode(SizeMode.MATCH_PARENT)
            .width(186 / 2f)
            .gravity(Gravity.CENTER_HORIZONTAL)
            .padding(0f, 33.5f, 0f, 116f)

        invPage.addChild("effect", effect)

        val wirelessPage = create(mainPos, true)
        wirelessPage.visibility = Widget.Visibility.GONE
        wirelessPage.isEnabled = false
        content.addChild("page_wireless", wirelessPage)

        val wirelessButton = createButton(Resource.Textures.ICON_WIRELESS)
        wirelessButton.layoutParams = WidgetContainer.LayoutParams()
            .widthMode(SizeMode.MATCH_PARENT)
            .height(16f)

        pageButtons.addChild("wireless", wirelessButton)
        pageButtons.onSelectionChanged = Consumer { button: RadioButtonWidget? ->
            when (button!!.name) {
                "inv" -> {
                    AnimationUtil.hide(wirelessPage)
                    AnimationUtil.show(invPage)
                    isHandleContainer = true
                    isRenderInventory = true
                }

                "wireless" -> {
                    AnimationUtil.hide(invPage)
                    AnimationUtil.show(wirelessPage)
                    isHandleContainer = false
                    isRenderInventory = false
                }
            }
        }
        pageButtons.selectButton(invButton)

        val info = create(this, (leftPos + imageWidth).toFloat(), (topPos - 22).toFloat())
        run {
            val p = WidgetContainer.LayoutParams()
                .gravity(Gravity.CENTER_RIGHT)
            val energyValueLabel = LabelWidget("0 AF")
            energyValueLabel.layoutParams = p
            energyValueSetter = { energyValueLabel.text = it }
            val energyLayout = createInfoRow("ENERGY", "icon_energy", -0xda3b01, energyValueLabel)
            info.addChild("energy_layout", energyLayout)

            val capacityValueLabel = LabelWidget("0 / 0")
            capacityValueLabel.layoutParams = p
            capacityValueSetter = { capacityValueLabel.text = it }
            val capacityLayout = createInfoRow("CAPACITY", "icon_capacity", -0x9400, capacityValueLabel)
            info.addChild("capacity_layout", capacityLayout)

            val infoLabel = LabelWidget("Information")
            infoLabel.layoutParams = LinearLayoutWidget.LayoutParams()
                .padding(6.5f, 0f, 0f, 0f)

            infoLabel.scale = 0.75f
            info.addChild("label_info", infoLabel)

            val rangeValueLabel = LabelWidget("0")
            rangeValueSetter = { rangeValueLabel.text = it }
            rangeValueLabel.layoutParams = WidgetContainer.LayoutParams()
                .height(LabelWidget.DEFAULT_BASE_FONT_SIZE)
                .gravity(Gravity.CENTER)

            val range = "Trans. Range"
            val rangeLayout = createAttributeRow(range, rangeValueLabel)
            info.addChild("range_layout", rangeLayout)

            val nameTextBox = TextBoxWidget(12)
            nameTextBox.background = null
            nameTextBox.setWhenEnter { s ->
                MisakaNetworkClient.send(
                    SetNodeNamePacket(wirelessNodeBlockEntity.blockPos, s)
                )
            }
            val name = "Node Name"
            val nameLayout = createAttributeRow(name, createInputRow(nameTextBox))
            info.addChild("name_layout", nameLayout)

            val passTextBox = TextBoxWidget(12)
            passTextBox.background = null
            passTextBox.setWhenEnter { s ->
                MisakaNetworkClient.send(
                    SetNodePassPacket(wirelessNodeBlockEntity.blockPos, s)
                )
            }
            val pass = "Password"
            val passLayout = createAttributeRow(pass, createInputRow(passTextBox))
            info.addChild("pass_layout", passLayout)
        }

        pageButtons.startAnimation(
            ObjectAnimator.ofFloat(
                { pageButtons.alpha = it }, 0f, 1f
            ).setDuration(childDuration)
        )
        pageButtons.startAnimation(
            ObjectAnimator.ofFloat(
                { pageButtons.translationY = it }, 20f, 0f
            ).setDuration(duration).setInterpolator(EasingFunctions.EASE_OUT_CUBIC)
        )

        updateInfo()
    }

    private fun updateInfo() {
        ticks++

        capacityValueSetter(wirelessNodeBlockEntity.connectedUsersCount.toString() + " / " + wirelessNodeBlockEntity.maxConnectedUsers)
        energyValueSetter(WindGenScreen.AF.format(wirelessNodeBlockEntity.energyStored))
        rangeValueSetter(wirelessNodeBlockEntity.radius.toString() + "")
    }

    override fun containerTick() {
        super.containerTick()
        updateInfo()
    }

    @SubscribeEvent
    fun onFocusGainedEvent(event: FocusGainedEvent) {
        isHandleContainer = false
    }

    @SubscribeEvent
    fun onFocusLostEvent(event: FocusLostEvent) {
        isHandleContainer = true
    }

    companion object {
        fun create(
            menu: WirelessNodeMenu,
            playerInventory: Inventory,
            title: Component,
            mainPos: BlockPos
        ): WirelessNodeScreen? {
            val level = Minecraft.getInstance().level
            val entity = level?.getBlockEntity(mainPos)
            return if (entity is WirelessNodeBlockEntity) {
                WirelessNodeScreen(menu, playerInventory, title, entity)
            } else {
                null
            }
        }
    }
}